lib/puma/launcher.rb

```diff
@@ -311,6 +313,32 @@ def prune_bundler
       end
     end
 
+    #
+    # Puma's systemd integration allows Puma to inform systemd:
+    #  1. when it has successfully started
+    #  2. when it is starting shutdown
+    #  3. periodically for a liveness check with a watchdog thread
+    #
+
+    def integrate_with_systemd
+      return unless ENV["SD_NOTIFY"]
+      return unless ENV["NOTIFY_SOCKET"]
+
+      require 'sd_notify'
+      return log "Systemd integration failed. It looks like you're trying to use systemd notify but don't have sd_notify gem installed" unless defined?(SdNotify)
+
+      require 'puma/systemd'
+
+      log "* Enabling systemd notification integration"
+
+      @events.on_booted { SdNotify.ready }
+      @events.on_stopped { SdNotify.stopping }
```

Comment:
Untested, but based on what https://github.com/sj26/puma-plugin-systemd does I think this also works:

```suggestion
      @events.on_booted { SdNotify.ready }
      @events.on_stopped { SdNotify.stopping }
      @events.on_restart { SdNotify.reloading }
```